ext/drbbdup/drbbdup.h

```diff
@@ -316,26 +324,27 @@ drbbdup_register_case_encoding(void *drbbdup_ctx, uintptr_t encoding);
 
 DR_EXPORT
 /**
- * Sets the runtime case encoding \p encoding.
+ * Sets the runtime case encoding \p encoding for a given thread.
```

Comment:
This is the complication with this approach: to change all the case encodings the client has to walk the threads and know where the TLS lives, and there's no function here to get the address easily.  It can be done via gencode but it's complex.  So drbbdup needs to provide the address to C code, or let the client control the TLS where it's stored so it can track the addresses on its own.  Certainly it would use its own thread TLS alloc list and atomics to write and avoid anything like dr_suspend_*.

I'm wondering if this is a better idea:

* Revert to the required insert_encode.
* Change drbbdup_get_encoding_opnd() to be a *register*.  Now the client can use a global var to hold the mode, and just do a load to the register at the top of every block.  It also has the flexibility (either b/c modes are not global, or b/c the barriers for prompt-enough shared loads on arm are too expensive) to keep the mode per-thread in its own TLS and load it to the register.  By having a register instead of memory there's no extra overhead.
* The only complication I see is drbbdup_set_encoding() for clients who want to set it from C code: since it could be far separated from the compare we wouldn't have it set the register in mcxt I don't think.  Either the client would have to implement it on its own by making a TLS slot and setting that and having insert_encode load from that TLS slot, or there's an option for drbbdup to load the value from its own TLS slot like it does today -- maybe if insert_encode is NULL it does that.

Reply:
I need to look further into to code - but on the top of my head, having drbbdup_get_encoding_opnd() be a register would require drbbdup to spill an additional GPR to handle dynamic case generation. Also, insert_encode may need to know which register is being reserved as it cannot be used for its own operation. For example, if it uses drreg to reserve a particular register and that register is already used by drbbdup, drreg will fail.

Reply:
Are you avoiding mem2mem stores because of register indirection and performance reasons?

Reply:
I don't want to have two memory refs when only one is necessary: give back a cache line to the app.

Reply:
Any interface that ends up with a single load, done atomically+visibly (so a barrier for ARM; load-acquire for AArch64; I plan to add helpers for such loads as part of #4215) so I can write to it from another thread, should work for me.  I would be ok with it having to be TLS if I can get the linear address for each thread and iterate over them; global would be simpler for my use case which has simple global phases -- not sure about other use cases.

Reply:
I don't think it is a good idea for the dispatcher to perform atomic operations. It needs to be fast.

Reply:
NVM, misread as saying internally, the dispatcher should load atomically.

Reply:
Okay so one problem is that preserving a register prior to the insert_encode limits the latter's flexibility. The use-case may require the original value of that register which is internally used by drbbdup, e.g., to compute an app address. I can inform the user that the register has the original app value, but I think the interface will be a bit messy. Storing the encoding into memory also allows easy integration with drreg - I can safely restore all registers after the insert_encode has been invoked without destroying the encoding and making dispatcher logic easier.

Reply:
I guess if a register is going to be used, then the user cannot reserve the same register in insert_encode using drreg. Drbbdup will reserve the register manually, so if the user is not careful and tries to reserve that register, drreg will return with success.

Reply:
Using memory instead of a register: what if the user could specify where the memory location is (pass an opnd_t)?

Reply:
Look at my last updated comment.

Regarding the specification of memory - that will result in drbbdup always performing an atomic load regardless of the use-case. I don't think that will help performance.

Reply:
Actually, I guess the user can pass a flag indicating whether to lock.

Reply:
To support dynamic fast path generation, I need guarantees that the memory passed will not change throughout the dispatching process - without the use of any locks. Can your use-case guarantee this?

Reply:
No my use case can't guarantee that: the trigger for a case changing is the global instruction count (this is #3995) and some other thread could cross the threshold and swap the case for all threads at any time.

I'm not sure what dynamic fast path generation is so I don't know whether it applies to my use case.  If not maybe there can be optional features for cases like mine that just swap between 2 global cases.

Reply:
I mean dynamic case generation - I use the terms interchangeably because ultimately one is generating different paths via drbbdup.

Reply:
Done (I think) but let me know if it can be better.

Reply:
Sorry, going to be busy next day or so, so may be a delay before I can take look.

Reply:
Don't worry, take your time.